File: backend/app/translator_protocol.py
from typing import Iterator, Protocol, runtime_checkable


@runtime_checkable
class Translator(Protocol):
    """翻译器统一接口。"""

    def translate(self, text: str) -> str:
        """同步翻译，返回译文。"""
        ...

    def translate_stream(self, text: str) -> Iterator[str]:
        """流式翻译，逐 token yield 译文片段。可选实现。"""
        ...

File: AutoTranslator/Sources/AutoTranslator/Translators/TranslatorProtocol.swift
import Foundation

protocol TranslatorProtocol: AnyObject {
    var source: String { get }
    var target: String { get }
    var supportsStreaming: Bool { get }
    func translate(_ text: String) async throws -> String
    func translateStream(_ text: String) -> AsyncThrowingStream<String, Error>
}

extension TranslatorProtocol {
    // Default: wrap non-streaming translate() in a single-yield stream.
    // Translators that natively support streaming should override.
    func translateStream(_ text: String) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    let result = try await self.translate(text)
                    continuation.yield(result)
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { @Sendable _ in task.cancel() }
        }
    }
}
